GrowthCouplingPotential

KristianJensen opened the pull request:
aka OptCouple
This PR adds my growth coupling potential based MILP optimization to the strain_design.deterministic.linear_programming module
I’ve also added some tests, but I don’t really know how pytest works, so I’ve just tried to mirror the OptKnock tests

Midnighter commented on tests/test_strain_design_deterministic.py:
```diff
@@ -146,3 +160,31 @@ def test_invalid_input(self, cplex_optknock):
+        _, growth_coupling_potential = gurobi_growth_coupling_potential
+        benchmark(growth_coupling_potential.run)
+
+    def test_result_is_correct(self, gurobi_growth_coupling_potential):
```
As far as I can tell, this test does not verify the function of the method. TPI is already essential for growth and the knock-out results do not modify that situation. (Please see the attached notebook.)
[opt_couple.ipynb.gz](https://github.com/biosustain/cameo/files/3378146/opt_couple.ipynb.gz)

Midnighter commented on cameo/strain_design/deterministic/linear_programming.py:
```diff
@@ -426,3 +430,394 @@ def _repr_html_(self):
+                    "%s has a forced flux. Remove the reaction or change its bounds for the function to work." % r.id)
+
+        # If a cobra.Reaction is given as target, convert to its ID string.
+        if isinstance(target, cobra.core.Reaction):
```
```suggestion
        if isinstance(target, cobra.Reaction):
```

Midnighter commented on cameo/strain_design/deterministic/linear_programming.py:
```diff
@@ -426,3 +430,394 @@ def _repr_html_(self):
+            knockout_reactions.append(target)
+
+        self.original_model = model
+        self.biomass_id = [r for r in model.reactions if r.objective_coefficient != 0][0].id
```
`objective_coefficient` is unfortunately deprecated. Maybe this could be inferred from the objective expression or specifically required as a parameter.

Midnighter commented on cameo/strain_design/deterministic/linear_programming.py:
```diff
@@ -426,3 +430,394 @@ def _repr_html_(self):
+        if remove_blocked:
+            blocked_reactions = cameo.flux_analysis.analysis.find_blocked_reactions(model)
+            model.remove_reactions(blocked_reactions)
+            blocked_reaction_ids = [r.id for r in blocked_reactions]
```
Turning this into a set will improve the following statements.

```suggestion
            blocked_reaction_ids = {r.id for r in blocked_reactions}
```

Midnighter commented on cameo/strain_design/deterministic/linear_programming.py:
```diff
@@ -426,3 +430,394 @@ def _repr_html_(self):
+            logger.debug("Removed " + str(len(blocked_reactions)) + " reactions that were blocked")
+
+        # Make dual
+        copied_model = model.copy()
```
Are the solver tolerances that were set above actually copied over?

KristianJensen replied:
I'm not sure, but that does not matter since the dual model is only constructed in order to copy stuff and add it back to the main model.

Midnighter commented on cameo/strain_design/deterministic/linear_programming.py:
```diff
@@ -426,3 +430,394 @@ def _repr_html_(self):
+
+        self.original_model = model
+        self.biomass_id = [r for r in model.reactions if r.objective_coefficient != 0][0].id
+        self.model = model = model.copy()
```
Assigning to `self.model` seems to have no effect since another model copy is assigned on line 691.
```suggestion
        model = model.copy()
```

Midnighter commented on cameo/strain_design/deterministic/linear_programming.py:
```diff
@@ -426,3 +430,394 @@ def _repr_html_(self):
+            problem.params.IntFeasTol = 1e-9
+            problem.params.MIPgapAbs = 1e-9
+            problem.params.MIPgap = 1e-9
+            # problem.params.TimeLimit = 200 # Use max 200 seconds when called, return best solution after that
```
What happened to these parameters? If they are not important, can these lines be removed?

Midnighter commented on cameo/strain_design/deterministic/linear_programming.py:
```diff
@@ -426,3 +430,394 @@ def _repr_html_(self):
+                1e-9)
+            problem.parameters.mip.tolerances.absmipgap.set(1e-9)
+            problem.parameters.mip.tolerances.mipgap.set(1e-9)
+            # problem.parameters.mip.pool.relgap.set(0.9)
```
Same question here.

Midnighter commented on cameo/strain_design/deterministic/linear_programming.py:
```diff
@@ -426,3 +430,394 @@ def _repr_html_(self):
+
+        # Fill the dictionaries with binary variables
+        # For the knockouts: only for the native reactions which are not exchanges
+        # for reaction in [r for r in model.reactions - excluded_reactions if r.type == "native"]:
```
```suggestion
```

Midnighter commented on cameo/strain_design/deterministic/linear_programming.py:
```diff
@@ -426,3 +430,394 @@ def _repr_html_(self):
+            interface = model.solver.interface
+            y_var = interface.Variable("y_" + reaction.id, type="binary")
+
+            # Constrain the primal: flux through reactions within (-1000, 1000)
```
This might depend on `cobra.Configuration.bounds`?

KristianJensen replied:
Yes, this should be taken into account now

Midnighter commented on cameo/strain_design/deterministic/linear_programming.py:
```diff
@@ -426,3 +430,394 @@ def _repr_html_(self):
+            interface = model.solver.interface
+            y_var = interface.Variable("y_" + reaction.id, type="binary")
+
+            # Constrain the primal: flux through reaction within (-1000, 1000)
```
Same question as above.

Midnighter commented on cameo/strain_design/deterministic/linear_programming.py:
```diff
@@ -426,3 +430,394 @@ def _repr_html_(self):
+            if "gurobi_interface" not in model.solver.interface.__name__:
+                raise NotImplementedError("Solution pools are currently only available with the Gurobi solver")
+
+            self.model.solver.problem.params.PoolSearchMode = 2
```
A comment here on what this means would be nice.
https://www.gurobi.com/documentation/8.1/refman/poolsearchmode.html

Midnighter commented on tests/test_strain_design_deterministic.py:
```diff
@@ -146,3 +160,31 @@ def test_invalid_input(self, cplex_optknock):
+    def test_growth_coupling_potential_runs(self, gurobi_growth_coupling_potential):
+        _, growth_coupling_potential = gurobi_growth_coupling_potential
+        result = growth_coupling_potential.run()
+        assert "obj_val" in result
```
This seems like it's not a very meaningful test. As long as there is no exception in the `run` method, these keys will always be present.

Midnighter commented on cameo/strain_design/deterministic/linear_programming.py:
```diff
@@ -426,3 +430,394 @@ def _repr_html_(self):
+    n_medium : int
+        The maximum allowed number of medium additions
+    use_solution_pool : Bool
+        Set to True when using the Gurobi solver to identify multiple alternative solutions at once
```
A reference to the published paper in the docstring would be very useful.

Midnighter commented on cameo/strain_design/deterministic/linear_programming.py:
```diff
@@ -426,3 +430,394 @@ def _repr_html_(self):
+        GrowthCouplingPotential(...).model.solver.problem
+
+
+    Parameters
```
```suggestion
    Attributes
```

bdelepine commented:
Hello!

Do you have any update on this PR? Is it still alive? Is it safe to use this branch to try out OptCouple?

KristianJensen commented:
Hi @bdelepine 

The PR is a bit stale as I haven't had time to finalize the unittests and incorporate the suggested changes. But the method should work. If you encounter any problems, feel free to post them here.

bdelepine commented on cameo/strain_design/deterministic/linear_programming.py:
```diff
@@ -426,3 +430,394 @@ def _repr_html_(self):
+    ----------
+    model : cobra.Model
+        A cobra model object containing the target reaction as well as all native and heterologous reactions.
+    target : cobra.Reaction or string
```
This differ from the signature of OptKnock class and its run method: 
```
OptKnock(model, exclude_reactions=None, remove_blocked=True, fraction_of_optimum=0.1, exclude_non_gene_reactions=True, use_nullspace_simplification=True, *args, **kwargs)
run(self, max_knockouts=5, biomass=None, target=None, max_results=1, *args, **kwargs)
```
An OptKnock object is created without mention of the target, but it is a parameter set in the `run()` method. Same remark for the number of knockouts. I suggest that it should be harmonized one way or the other in both classes, I guess keeping OptKnock strategy for backward compatibility.

KristianJensen replied:
Thanks for pointing that out. Regarding the number of knockouts etc it should be relatively easy to change to follow the OptKnock signature. The target is a bit more tricky. The reason I put it as an argument to `__init__` is that the choice of target reaction affects the construction of the optimization problem more than is the case for OptKnock. I will think about whether/how this can be changed

bdelepine replied:
Honestly, it might be better for cameo to accept that the signature between optimisation classes (and there associated run method)  is **expected** to be different for each class. This would give more flexibility to create the best preprocessing environment for the problem, just like you did. However this is far from an "object-oriented" mindset, which can be surprising for users. So, an acceptable solution IMHO would be to simply warn users in cameo documentation, but really, this is a design decision that should be made by package maintainers since the question will arise for each new algorithm implemented in cameo.

bdelepine commented:
Hi @KristianJensen,

Sadly, I confirm that the current branch does not work. The `sol.f` triggered an AttributeError. Once corrected, the code runs (using CPLEX). Unfortunately, I was still unable to get results since I keep getting `{'knockouts': [], 'knockins': [], 'medium': [], 'obj_val': 0.0}` despite trying different set of parameters. I guess my example is broken or that something is silently off somewhere 😕.

HTH, +

KristianJensen commented:
Thanks for the report @bdelepine 
I will try to get this PR finished ASAP - I will also need to use it myself in my current work.
I have now addressed some of @Midnighter's comments and updated the code to be compatible with modern cobrapy. I will do some testing of the method next.

If you want to try the method in the meantime, there is an implementation in the supplementary material of the paper, including a list of package versions.

KristianJensen commented:
@Midnighter I have pushed some changes:
- The code is now compatible with modern cobrapy
- The test no longer relies on gurobi, but runs with glpk
- I've changed the test to a target that is not already growth-coupled (and added an assertion to verify this)

Let me know if there are other issues I should address (or if I should ping someone else instead :) )

carrascomj reviewed (approved):
Pinging requester @KristianJensen 
I merged latest changes into this branch, CI is passing and I think is ready for merging into devel.

phantomas1234 commented:
Thanks @carrascomj!
